### ISSUE TITLE
Enhancement for p:directory-list

### ISSUE THREAD

**Issue by xml-project** [xproc/3.0-specification#234]
Working on a recent example I find it very inconvenient that the options "include-filter" and "exclude-filter" work an files and directories likewise because I want all files matching a pattern and all directories whatever name they have.

I think that is a typical use case so I propose to add an option "apply-to-directories" which takes xs:boolean as values. If the option is true, filters are applied to directory names, if it is false, every found directory is included in the result. The new option is NOT required. Its default value is "true()".

**Comment by ndw** [comment]
Have you considered how this interacts with globbing: http://spec.xproc.org/master/head/steps30/#uri-globbing

I was sort of hoping that globbing would more-or-less replace the include/exclude filters in the common cases.

**Comment by xml-project** [comment]
Nope, sorry. I am just fighting with an XProc 1.0 pipeline (which needs to be short because I want to put it on a slide).

I agree, that it will replace exclude/include filters in most cases, but at the moment I can not see how it would help with "the directories all, files only if they match"-case since directory-names are not represented in a special but in a special element. What do I miss?

**Comment by gimsieke** [comment]
A recursive directory listing that applies its include or exclude filters to directories and files likewise will only be useful in practice if it is used without the filters.
The proposed boolean option `apply-to-directories` will only be useful in the non-recursive case. If you want to extend `p:directory-list` to recursive operation (which I think is desirable), you either need include and exclude filters for directories and files separately (makes 4 filter attributes), or you can probably address this with globbing. 
Achim, can’t the requirement “all directories, only files matching 'IMG_*.jpg'” be represented by a glob pattern `/**/IMG_*.jpg`? 
Is your use case different from this?

**Comment by xml-project** [comment]
I am sorry, but obviously I do not understand the concept of a glob. To my reading the glob in p:directory-list is applied to $path, that is, which directory is listed. The two filter-options are marked as regex and there is a reference to "regex" as explain in XPath/XQuery-functions.
The problem I would like to address applies to  the content of the selected directory, that is: Which entries are listed. The filter I can define here are regex applied to the name of the content of the selected directory, irrespective
of the type (file or directory).
@gimsieke 
1. I agree with the idea of making p:directory-list recursive as a library step.
2. Given that globbing is possible in "include-filter" (to my reading currently is it not) and it p:directory-list would be recursive, your proposal would select all "IMG_*.jpg" in all folders, so the result will be a lot of "c:file" elements, but no "c:directory"-elements, because they typically have no have a name matching the pattern. So if I want to find all files matching a certain glob irrespective of the directory, sub-directory etc. I am fine with your proposal. But as I said, if I want a list of all the directories to, I can not do this with one p:directory-list.

**Comment by gimsieke** [comment]
Maybe we should clarify that “path” in the globbing section is not just for the directory part, but for the whole fully qualified file name, i.e., directory path plus file base name, where base name here includes any extensions. This becomes implicitly clear from the examples that Norm gave. 
Does it make sense if you interpret path like this?

**Comment by gimsieke** [comment]
> Given that globbing is possible in "include-filter" (to my reading currently is it not) and it p:directory-list would be recursive, your proposal would select all "IMG_*.jpg" in all folders, so the result will be a lot of "c:file" elements, but no "c:directory"-elements, because they typically have no have a name matching the pattern. So if I want to find all files matching a certain glob irrespective of the directory, sub-directory etc. I am fine with your proposal. But as I said, if I want a list of all the directories to, I can not do this with one p:directory-list.

Ok, you are addressing some more real issues.

First I think that `p:directory-list` still has a `path` attribute that is not a glob, but a file uri (apart from implementation-defined other supported URI schemes, or after catalog resolution which would also only happen implementation-definedly).
This `$path` sets the base for the step.
I don’t think that we can use glob patterns in `$path` because otherwise the directory listing always will start at the file system root, but users don’t want all the ancestor paths, they want a listing of a given directory.
Then we can replace `include-filter` (and maybe `exclude-filter` – which important exclusion use cases will `exclude-filter` fall short of supporting?) with a single `glob` option that specifies patterns for both, subdirectories and file names. 
Glob patterns need to be anchored. I’d suggest that if they are given for `p:directory-list`, they should be interpreted as anchored to the subdirectory given in `path`. That is, both `/foo` and `foo` match `file:/my/base/dir/foo` if `path` is `file:/my/base/dir/` or `file:/my/base/dir` (trailing slash normalization should be performed on `path`).

Secondly, I think that `p:directory-list` should always return a structure with nested directories, not just flat `c:file` entries, if the glob pattern matches files down the subdir tree and if `$recurse = true()`.
If the glob pattern is like `*/*.jpg` and recurse is `false()`, I don’t think that there should be anything else below the top-level `/c:directory` because the files, if present, would be in a subdirectory of `path`.

**Comment by ndw** [comment]
1. Yes, I intended globbing to apply to the whole path+filename, not just the path.
2. We clearly need a more detailed spec for what `p:directory-list` does.
3. I'm not sure I understand all the proposed requirements.

**Comment by xml-project** [comment]
Thanks @ndw 
Proposal: Close this issue and open a new one "Detail specs for p:directory-list" with reference to this discussion.
Agreed?

**Comment by ndw** [comment]
Sure...I'm just trying to write up some of the requirements now. I don't feel strongly about whether we leave them in this issue or start a new one.

BTW, in the context of a URI, I think "path" includes the filename:

```
scheme:[//[user[:password]@]host[:port]][/path][?query][#fragment]
```

**Comment by xml-project** [comment]
Ok, so let us leave the discussion here. Enhancement for p:directory-list seems to fit anyway.

**Comment by ndw** [comment]
@xml-project I'm not sure I understand the requirement you expressed in the issue description.
Suppose you have:

```
/a/b/file.txt
/a/c/image.jpg
```

If you search for JPG images, starting at `/a/`, are you saying you want `/a/b` in the result even though that directory doesn't contain any JPG images?

**Comment by gimsieke** [comment]
Re-visiting the proposed glob spec – it is really meant to capture a file path from front to back. 

Let’s look at an example that Achim provided. There’s a directory structure
```
file:///users/berndzen/documents/ordner1
  - document1.xml
  - document2.xml
  - Ordner-2/
  - grafik1.jpg
  - grafik2.jpg
  - Directory-4/
```
Desired result:
```xml
<c:directory name="ordner1" xml:base="file:///users/berndzen/documents/ordner1" 
  xmlns:c="http://www.w3.org/ns/xproc-step">
  <c:file name="document1.xml" />
  <c:file name="document2.xml" />
  <c:directory name="Ordner-2" />
  <c:directory name="Directory-4" />
</c:directory>
```

Suppose that the `path` option accept globs. Then we can us as glob: `file:///users/berndzen/documents/ordner1/[^g]*`

But what if the folder 'ordner1' contained a file 'groups.xml' that should be included in the result. Then we need to think about a glob pattern that matches '*.xml' and also files/folders without an extension. Regexes surely give more flexibility in specifying alternatives or optional path / filename parts.

And I think that if recursive mode is an option, we might find it more convenient to supply separate regexes for files and folders than to always supply a pattern for a whole path. 

These are certainly two different use cases: 1. Get me all `*.jpg` no matter how deeply nested, with all intermediate directories. 2. Give me a complete subtree listing, but ignore `.svn` subdirs and `*~` files.

Some of the use cases can be supported by applying one or more `p:delete` after a complete listing.

So my proposal is to keep the path option as a fixed URI and have separate inclusion and exclusion regexes for both files and folder names. 

Alternatively, what Achim proposed, an option not to apply the inclusion pattern to folders. But I’d prefer a default include pattern of '.*' for both files and folders so we don’t need this option.

I don’t know whether globs are useful in other places like `p:load`, but here I’d prefer to do without them because we have regexes.

**Comment by xml-project** [comment]
> Alternatively, what Achim proposed, an option not to apply the inclusion pattern to folders. But I’d prefer a default include pattern of '.*' for both files and folders so we don’t need this option.

Yes, if we have different patterns for files and folder we do not need the proposed option. And we are more flexible because we can include/exclude certain directories by matching there name.
